### PR TITLE
Enabling an Umbraco admin user to disable 2FA for a member

### PR DESCRIPTION
**Pull request by elit0451**
## Details
- Adding the possibility for admins to disable Two-Factor Authentication for a specific member, in case the member has no longer access to their authenticator app;
- New Two-Factor Authentication toggle has been added to the Member content app in the Members section;
- ❗ 2FA can **only** be **disabled** using the UI. If you want to enable it, you still need to go through the process of setting it up ([_ref. docs_](https://our.umbraco.com/documentation/reference/security/two-factor-authentication/#two-factor-authentication-for-members)). 

- Create a member in the backoffice;
</br>

> Setup 2FA for Members
- Install the [GoogleAuthenticator NuGet Package](https://www.nuget.org/packages/GoogleAuthenticator/) to _Umbraco.Web.UI_ project;
  - _It will work for both Google and Microsoft authenticator apps_; 
- Create `UmbracoAppAuthenticator.cs` class _(Check Appendix)_;
- Add the following lines in `Startup.cs`:
```c#
using Umbraco.Cms.Core.Security;
using Umbraco.Cms.Web.BackOffice.Security;

. . .
        public void ConfigureServices(IServiceCollection services)
        {
            var memberIdentityBuilder = new MemberIdentityBuilder(services);
            memberIdentityBuilder.AddTwoFactorProvider<UmbracoAppAuthenticator>(UmbracoAppAuthenticator.Name);
            
            services.AddUmbraco(_env, _config)
              . . .
        }   
. . .
```
- Create the following **Partial View Macro Files** from snippets;
  - LoginStatus.cshtml;
  - Login.cshtml (_there are changes in the file, so copy the code from Appendix_);
  - EditProfile.cshtml (_there are changes in the file, so copy the code from Appendix_);
- Enable _"Use in rich text editor and the grid"_ in their corresponding Macros;
- Create a Doc type with Richtext editor;
- Display the value of the Richtext editor in the Template of that doc type;
- Create a content item and add the Macros in the following order:
  - Login;
  - EditProfile;
  - LoginStatus; 
</br>

> Enable 2FA for Members
- Verify that the Two-Factor Authentication toggle (in the Member section > Member content app) is turned off & can't be enabled;
- Visit the website's initial page and login with the member account you created;
- Setup UmbracoAppAuthenticator by following the instructions;
- Logout and login again and try out the 2FA for member;
- Go back to the Two-Factor Authentication toggle and verify that it has been enabled;
- Verify that there is an entry for that member in the `[umbracoTwoFactorLogin]` table;
  - You can easily compare the guid from the url with the value in the `[userOrMemberKey]`column; 😉 
</br>

> Disable 2FA for Members
- Toggle off the Two-Factor Authentication toggle & Save;
- Go back to the `[umbracoTwoFactorLogin]` table and verify that the entry for that member has been deleted;
- Try to login as that member and verify that the 2FA setup is displayed once again.

## Appendix
<details><summary><b>Contents of <code>UmbracoAppAuthenticator.cs</code></b></summary>
</br>

```c#
using System;
using System.Threading.Tasks;
using Google.Authenticator;
using Umbraco.Cms.Core.Security;
using Umbraco.Cms.Core.Services;

namespace Umbraco.Cms.Web.UI
{
    /// <summary>
    /// Model with the required data to setup the authentication app.
    /// </summary>
    public class QrCodeSetupData
    {
        /// <summary>
        /// The secret unique code for the user and this ITwoFactorProvider.
        /// </summary>
        public string? Secret { get; init; }

        /// <summary>
        /// The SetupCode from the GoogleAuthenticator code.
        /// </summary>
        public SetupCode? SetupCode { get; init; }
    }

    /// <summary>
    /// App Authenticator implementation of the ITwoFactorProvider
    /// </summary>
    public class UmbracoAppAuthenticator : ITwoFactorProvider
    {
        /// <summary>
        /// The unique name of the ITwoFactorProvider. This is saved in a constant for reusability.
        /// </summary>
        public const string Name = "UmbracoAppAuthenticator";

        private readonly IMemberService _memberService;

        /// <summary>
        /// Initializes a new instance of the <see cref="UmbracoAppAuthenticator"/> class.
        /// </summary>
        public UmbracoAppAuthenticator(IMemberService memberService)
        {
            _memberService = memberService;
        }

        /// <summary>
        /// The unique provider name of ITwoFactorProvider implementation.
        /// </summary>
        /// <remarks>
        /// This value will be saved in the database to connect the member with this  ITwoFactorProvider.
        /// </remarks>
        public string ProviderName => Name;

        /// <summary>
        /// Returns the required data to setup this specific ITwoFactorProvider implementation. In this case it will contain the url to the QR-Code and the secret.
        /// </summary>
        /// <param name="userOrMemberKey">The key of the user or member</param>
        /// <param name="secret">The secret that ensures only this user can connect to the authenticator app</param>
        /// <returns>The required data to setup the authenticator app</returns>
        public Task<object> GetSetupDataAsync(Guid userOrMemberKey, string secret)
        {
            var member = _memberService.GetByKey(userOrMemberKey);

            var twoFactorAuthenticator = new TwoFactorAuthenticator();
            SetupCode? setupInfo = null;

            if (member is not null)
            {
                setupInfo = twoFactorAuthenticator.GenerateSetupCode("My application name", member.Username, secret, false);
            }

            return Task.FromResult<object>(new QrCodeSetupData()
            {
                SetupCode = setupInfo,
                Secret = secret
            });
        }

        /// <summary>
        /// Validated the code and the secret of the user.
        /// </summary>
        public bool ValidateTwoFactorPIN(string secret, string code)
        {
            var twoFactorAuthenticator = new TwoFactorAuthenticator();
            return twoFactorAuthenticator.ValidateTwoFactorPIN(secret, code);
        }

        /// <summary>
        /// Validated the two factor setup
        /// </summary>
        /// <remarks>Called to confirm the setup of two factor on the user. In this case we confirm in the same way as we login by validating the PIN.</remarks>
        public bool ValidateTwoFactorSetup(string secret, string token) => ValidateTwoFactorPIN(secret, token);
    }
}

```

</details>

<details><summary><b>Contents of <code>Login.cshtml</code></b></summary>
</br>

```c#
@inherits Umbraco.Cms.Web.Common.Macros.PartialViewMacroPage


@using Umbraco.Cms.Web.Common.Models
@using Umbraco.Cms.Web.Common.Security
@using Umbraco.Cms.Web.Website.Controllers
@using Umbraco.Cms.Core.Services
@using Umbraco.Extensions
@inject IMemberExternalLoginProviders memberExternalLoginProviders
@inject ITwoFactorLoginService twoFactorLoginService
@{
    var loginModel = new LoginModel();
    // You can modify this to redirect to a different URL instead of the current one
    loginModel.RedirectUrl = null;
    
    var isLoggedIn = Context.User?.Identity?.IsAuthenticated ?? false;
}

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validate/1.19.0/jquery.validate.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validation-unobtrusive/3.2.11/jquery.validate.unobtrusive.min.js"></script>


@if (!isLoggedIn)
{
@if (ViewData.TryGetTwoFactorProviderNames(out var providerNames))
{

    foreach (var providerName in providerNames)
    {
        <div class="2fa-form">
            <h4>Two factor with @providerName.</h4>
            <div asp-validation-summary="All" class="text-danger"></div>
            @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.Verify2FACode)))
            {

                <text>
                    <input type="hidden" name="provider" value="@providerName"/>
                    Input security code: <input name="code" value=""/><br/>
                    <button type="submit" class="btn btn-primary">Validate</button>
                    <div asp-validation-summary="ModelOnly" class="text-danger"></div>
                </text>
            }
        </div>
    }

}
else
{


<div class="login-form">

    @using (Html.BeginUmbracoForm<UmbLoginController>(
        "HandleLogin", new { RedirectUrl = loginModel.RedirectUrl })) {

            <h4>Log in with a local account.</h4>
            <hr />
            <div asp-validation-summary="ModelOnly" class="text-danger"></div>
            <div class="mb-3">
                <label asp-for="@loginModel.Username" class="form-label"></label>
                <input asp-for="@loginModel.Username" class="form-control" />
                <span asp-validation-for="@loginModel.Username" class="form-text text-danger"></span>
            </div>
            <div class="mb-3">
                <label asp-for="@loginModel.Password" class="form-label"></label>
                <input asp-for="@loginModel.Password" class="form-control" />
                <span asp-validation-for="@loginModel.Password" class="form-text text-danger"></span>
            </div>
            <div class="mb-3 form-check">
                <input asp-for="@loginModel.RememberMe" class="form-check-input">
                <label asp-for="@loginModel.RememberMe" class="form-check-label">
                    @Html.DisplayNameFor(m => loginModel.RememberMe)
                </label>
            </div>

            <button type="submit" class="btn btn-primary">Log in</button>


        }
@{
    var loginProviders = await memberExternalLoginProviders.GetMemberProvidersAsync();
    var externalSignInError = ViewData.GetExternalSignInProviderErrors();

    if (loginProviders.Any())
    {
        <hr/>
        <h4>Or using external providers</h4>
        if (externalSignInError?.AuthenticationType is null && externalSignInError?.Errors.Any() == true)
        {
            @Html.DisplayFor(x => externalSignInError.Errors);
        }

        @foreach (var login in await memberExternalLoginProviders.GetMemberProvidersAsync())
        {

            @using (Html.BeginUmbracoForm<UmbExternalLoginController>(nameof(UmbExternalLoginController.ExternalLogin)))
            {
                <button type="submit" name="provider" value="@login.ExternalLoginProvider.AuthenticationType">
                    Sign in with @login.AuthenticationScheme.DisplayName
                </button>

                if (externalSignInError?.AuthenticationType == login.ExternalLoginProvider.AuthenticationType)
                {
                    @Html.DisplayFor(x => externalSignInError.Errors);
                }
            }
        }
    }
}
</div>
}
}

```

</details>

<details><summary><b>Contents of <code>EditProfile.cshtml</code></b></summary>
</br>

```c#
@inherits Umbraco.Cms.Web.Common.Macros.PartialViewMacroPage

@using Umbraco.Cms.Core.Services
@using Umbraco.Cms.Web.Common.Security
@using Umbraco.Cms.Web.Website.Controllers
@using Umbraco.Cms.Web.Website.Models
@using Umbraco.Extensions
@inject MemberModelBuilderFactory memberModelBuilderFactory;
@inject IMemberExternalLoginProviders memberExternalLoginProviders
@inject IExternalLoginWithKeyService externalLoginWithKeyService
@inject ITwoFactorLoginService twoFactorLoginService
@{

    // Build a profile model to edit
    var profileModel = await memberModelBuilderFactory
        .CreateProfileModel()
        .BuildForCurrentMemberAsync();

    var success = TempData["FormSuccess"] != null;

    var loginProviders = await memberExternalLoginProviders.GetMemberProvidersAsync();
    var externalSignInError = ViewData.GetExternalSignInProviderErrors();

    var currentExternalLogin = profileModel is null
        ? new Dictionary<string, string>()
        : externalLoginWithKeyService.GetExternalLogins(profileModel.Key).ToDictionary(x => x.LoginProvider, x => x.ProviderKey);
}

<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validate/1.19.0/jquery.validate.min.js"></script>
<script src="https://cdnjs.cloudflare.com/ajax/libs/jquery-validation-unobtrusive/3.2.11/jquery.validate.unobtrusive.min.js"></script>

@if (profileModel != null)
{
    if (success)
    {
        @* This message will show if profileModel.RedirectUrl is not defined (default) *@
        <p class="text-success">Profile updated</p>
    }

    using (Html.BeginUmbracoForm<UmbProfileController>("HandleUpdateProfile", new { RedirectUrl = profileModel.RedirectUrl }))
    {
        <h2>Update your account.</h2>
        <hr />
        <div asp-validation-summary="All" class="text-danger"></div>
        <div class="mb-3">
            <label asp-for="@profileModel.Name" class="form-label"></label>
            <input asp-for="@profileModel.Name" class="form-control" aria-required="true" />
            <span asp-validation-for="@profileModel.Name" class="form-text text-danger"></span>
        </div>
        <div class="mb-3">
            <label asp-for="@profileModel.Email" class="form-label"></label>
            <input asp-for="@profileModel.Email" class="form-control" autocomplete="username" aria-required="true" />
            <span asp-validation-for="@profileModel.Email" class="form-text text-danger"></span>
        </div>

        @if (!string.IsNullOrWhiteSpace(profileModel.UserName))
        {
            <div class="mb-3">
                <label asp-for="@profileModel.UserName" class="form-label"></label>
                <input asp-for="@profileModel.UserName" class="form-control" autocomplete="username" aria-required="true" />
                <span asp-validation-for="@profileModel.UserName" class="form-text text-danger"></span>
            </div>
        }

        @if (profileModel.MemberProperties != null)
        {
            for (var i = 0; i < profileModel.MemberProperties.Count; i++)
            {
                <div class="mb-3">
                    @Html.LabelFor(m => profileModel.MemberProperties[i].Value, profileModel.MemberProperties[i].Name)
                    <input asp-for="@profileModel.MemberProperties[i].Value" class="form-control" />
                    @Html.HiddenFor(m => profileModel.MemberProperties[i].Alias)
                    <span asp-validation-for="@profileModel.MemberProperties[i].Value" class="form-text text-danger"></span>
                </div>
            }
        }

        <button type="submit" class="btn btn-primary">Update</button>

        if (loginProviders.Any())
        {
            <hr />
            <h4>Link external accounts</h4>

            if (externalSignInError?.AuthenticationType is null && externalSignInError?.Errors.Any() == true)
            {
                @Html.DisplayFor(x => externalSignInError.Errors);
            }

            @foreach (var login in loginProviders)
            {
                if (currentExternalLogin.TryGetValue(login.ExternalLoginProvider.AuthenticationType, out var providerKey))
                {
                    @using (Html.BeginUmbracoForm<UmbExternalLoginController>(nameof(UmbExternalLoginController.Disassociate)))
                    {
                        <input type="hidden" name="providerKey" value="@providerKey" />
                        <button type="submit" name="provider" value="@login.ExternalLoginProvider.AuthenticationType">
                            Un-Link your @login.AuthenticationScheme.DisplayName account
                        </button>

                        if (externalSignInError?.AuthenticationType == login.ExternalLoginProvider.AuthenticationType)
                        {
                            @Html.DisplayFor(x => externalSignInError.Errors);
                        }
                    }
                }
                else
                {
                    @using (Html.BeginUmbracoForm<UmbExternalLoginController>(nameof(UmbExternalLoginController.LinkLogin)))
                    {
                        <button type="submit" name="provider" value="@login.ExternalLoginProvider.AuthenticationType">
                            Link your @login.AuthenticationScheme.DisplayName account
                        </button>

                        if (externalSignInError?.AuthenticationType == login.ExternalLoginProvider.AuthenticationType)
                        {
                            @Html.DisplayFor(x => externalSignInError.Errors);
                        }
                    }
                }

            }
        }
    }
}
@{
    <br />
    // Show all two factor providers
    var providerNames = twoFactorLoginService.GetAllProviderNames();
    if (providerNames.Any())
    {
        <div asp-validation-summary="All" class="text-danger"></div>
        foreach (var providerName in providerNames)
        {
            object setupData = null;

            if (profileModel is not null)
            {
                setupData = await twoFactorLoginService.GetSetupInfoAsync(profileModel.Key, providerName);
            }

            var isLoggedIn = Context.User?.Identity?.IsAuthenticated ?? false;

            if (setupData is null && isLoggedIn)
            {
                @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.Disable)))
                {
                    <input type="hidden" name="providerName" value="@providerName" />
                    <button type="submit">Disable @providerName</button>
                }

            }
            else if (setupData is QrCodeSetupData qrCodeSetupData)
            {
                @using (Html.BeginUmbracoForm<UmbTwoFactorLoginController>(nameof(UmbTwoFactorLoginController.ValidateAndSaveSetup)))
                {
                    <h3>Setup @providerName</h3>
                    <img src="@qrCodeSetupData.SetupCode.QrCodeSetupImageUrl" />
                    <p>Scan the code above with your authenticator app <br /> and enter the resulting code here to validate:</p>
                    <input type="hidden" name="providerName" value="@providerName" />
                    <input type="hidden" name="secret" value="@qrCodeSetupData.Secret" />
                    <input type="text" name="code" />
                    <button type="submit">Validate & save</button>
                }

            }
        }
    }
}

```

</details>

### CODE REVIEW

File: src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs [hunk 1]
```diff
@@ -46,7 +54,32 @@ public class MemberTabsAndPropertiesMapper : TabsAndPropertiesMapper<IMember>
         _memberService = memberService ?? throw new ArgumentNullException(nameof(memberService));
         _memberGroupService = memberGroupService ?? throw new ArgumentNullException(nameof(memberGroupService));
         _memberPasswordConfiguration = memberPasswordConfiguration.Value;
-        _propertyEditorCollection = propertyEditorCollection;
+        _twoFactorLoginService = twoFactorLoginService ?? throw new ArgumentNullException(nameof(twoFactorLoginService));
+    }
+
+    [Obsolete("Use other ctor - Will be removed in Umbraco 13")]
```
Comment (Zeegaan) [inline src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs:60]: We should explain what the other constructor is to be helpful 😁 
So should be `Use constructor that also takes a ITwoFactorLoginService, this is scheduled for removal in V13 `

File: src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs [hunk 2]
```diff
@@ -26,8 +28,13 @@ public class MemberTabsAndPropertiesMapper : TabsAndPropertiesMapper<IMember>
     private readonly IMemberService _memberService;
     private readonly IMemberGroupService _memberGroupService;
     private readonly MemberPasswordConfigurationSettings _memberPasswordConfiguration;
-    private readonly PropertyEditorCollection _propertyEditorCollection;
+    private readonly ITwoFactorLoginService _twoFactorLoginService;
```
Comment (Zeegaan) [inline src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs:31]: Might be out of scope entirely, but I really do not like services in mappers, in theory, they should just be mapping 1:1, and if services are needed, we should use a factory instead 🤔 
This is of course major refactoring so will take a bit of time, so might be a great topic for NTT to discuss how we want to move forward👍

File: src/Umbraco.Web.BackOffice/Controllers/MemberController.cs [hunk 3]
```diff
@@ -101,6 +107,44 @@ public class MemberController : ContentControllerBase
         _shortStringHelper = shortStringHelper;
         _passwordChanger = passwordChanger;
         _scopeProvider = scopeProvider;
+        _twoFactorLoginService = twoFactorLoginService;
+    }
+
+    [Obsolete("Use other ctor - Will be removed in Umbraco 13")]
```
Comment (Zeegaan) [inline src/Umbraco.Web.BackOffice/Controllers/MemberController.cs:113]:
```suggestion
    [Obsolete("Use constructor that also takes an ITwoFactorLoginService instead, this is scheduled for removal in Umbraco 13")]
```

**Review by Zeegaan — CHANGES_REQUESTED** [review]
Have a few comments before i get to testing 😁

File: src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs (hunk 2 above)
Comment (elit0451) [reply src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs:31]: Yeah, let's bring that up on Monday 😁

File: src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs (hunk 1 above)
Comment (elit0451) [reply src/Umbraco.Core/Models/Mapping/MemberTabsAndPropertiesMapper.cs:60]: I will be more concrete then 👍

**Comment by nul800sebastiaan** [comment]
Cherry picked for 10.4.0 dabc7bed9b06065dbd4e3ddb759acc9c7084894c

**Pull request merged**